server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -289,36 +293,210 @@ static class Compatible extends FiltersAggregator {
+
+        final DocIdSetIterator[] filteredDocIdSetIterators;
+        final int[] lastMatches;
+        final DocId currentDocID;
+        final int maxDocId;
+
+        // Tracks next doc matches as they get returned from the per-filter iterators.
+        final PriorityQueue<Integer> nextDocs;
+
+        CompetitiveIterator(DocIdSetIterator[] filteredDocIdSetIterators, int[] lastMatches, DocId currentDocID, int maxDocId) {
+            this.filteredDocIdSetIterators = filteredDocIdSetIterators;
+            this.lastMatches = lastMatches;
+            this.currentDocID = currentDocID;
+            this.maxDocId = maxDocId;
+
+            nextDocs = new PriorityQueue<>(filteredDocIdSetIterators.length) {
+                @Override
+                public boolean offer(Integer e) {
+                    // Skip inserting duplicate values.
+                    if (contains(e)) {
+                        return false;
+                    }
+                    return super.offer(e);
+                }
+            };
+        }
+
+        /**
+         * Returns the next possible filter match, starting from the passed target doc.
+         * It uses the priority queue to determine if it can skip ranges with no matches.
+         */
+        private int getNextDoc(int target) {
+            if (nextDocs.isEmpty()) {
+                return target;
+            }
+
+            // If the next doc in the priority queue is further ahead, skip the rest.
+            int nextDoc = nextDocs.poll();
+            if (nextDoc > target) {
+                return nextDoc;
+            }
+
+            // There's at least one doc in the priority queue that's not further ahead so we can't skip
+            // any docs. Remove any past docs from the priority queue, not needed any more.
+            while (nextDocs.isEmpty() == false && nextDocs.peek() <= target) {
+                nextDocs.poll();
+            }
+            return target;
+        }
+
+        @Override
+        public int docID() {
+            return currentDocID.get();
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return advance(currentDocID.get() + 1);
+        }
+
+        @Override
+        public int advance(int target) throws IOException {
+            target = getNextDoc(target);
+            if (target >= maxDocId) {
+                target = DocIdSetIterator.NO_MORE_DOCS;
+                currentDocID.set(target);
+                return target;
+            }
+            for (int i = 0; i < filteredDocIdSetIterators.length; i++) {
+                if (lastMatches[i] < target) {
+                    lastMatches[i] = filteredDocIdSetIterators[i].advance(target);
+                    nextDocs.add(lastMatches[i]);
+                }
+            }
+            currentDocID.set(target);
+            return target;
+        }
+
+        @Override
+        public long cost() {
+            return maxDocId;
```

Comment:
I think this should return the filteredDocIdSetIterators's cost with highest cost.

Reply:
Not needed any more.